src/dex/swaap-v2/swaap-v2.ts

```diff
@@ -0,0 +1,713 @@
+                sender: 'address',
+                fromInternalBalance: 'bool',
+                recipient: 'address',
+                toInternalBalance: 'bool',
+              },
+              limit: 'uint256',
+            },
+          },
+        ],
+        [
+          isBatchSwap,
+          callerSlot,
+          {
+            swap: {
+              poolId: callDataStruct.singleSwap.poolId,
+              kind: callDataStruct.singleSwap.kind,
+              assetIn: callDataStruct.singleSwap.assetIn,
+              assetOut: callDataStruct.singleSwap.assetOut,
+              amount: callDataStruct.singleSwap.amount,
+              userData: callDataStruct.singleSwap.userData,
+            },
+            funds: {
+              sender: callDataStruct.funds.sender,
+              fromInternalBalance: callDataStruct.funds.fromInternalBalance,
+              recipient: callDataStruct.funds.recipient,
+              toInternalBalance: callDataStruct.funds.toInternalBalance,
+            },
+            limit: side === SwapSide.BUY ? MAX_UINT : '0',
+          },
+        ],
+      );
+    }
+
+    return {
+      targetExchange: router,
+      payload: payload,
+      networkFee: '0',
+    };
+  }
+
+  async getSimpleParam(
+    srcToken: string,
+    destToken: string,
+    srcAmount: string,
+    destAmount: string,
+    data: SwaapV2Data,
+    side: SwapSide,
+  ): Promise<SimpleExchangeParam> {
+    const { router, callData } = data;
+
+    assert(
+      router !== undefined,
+      `${this.dexKey}-${this.network}: router undefined`,
+    );
+
+    assert(
+      callData !== undefined,
+      `${this.dexKey}-${this.network}: callData undefined`,
+    );
+
+    return this.buildSimpleParamWithoutWETHConversion(
+      srcToken,
+      srcAmount,
+      destToken,
+      destAmount,
+      callData,
+      router,
+    );
+  }
+
+  computeMaxLiquidity = (
+    levels: SwaapV2PriceLevels,
+    tokenAddress: string,
+  ): number => {
+    if (tokenAddress === levels.base!) {
+      return (
+        (levels.asks[levels.asks.length - 1]?.level ?? 0) *
+        levels.asks[0]?.price
+      );
+    }
+    return (
+      (levels.asks[levels.bids.length - 1]?.level ?? 0) * levels.asks[0]?.price
+    );
+  };
+
+  async getTopPoolsForToken(
+    tokenAddress: Address,
+    limit: number,
+  ): Promise<PoolLiquidity[]> {
+    const normalizedTokenAddress = normalizeTokenAddress(tokenAddress);
+
+    const pLevels = await this.getCachedLevels();
+
+    if (pLevels === null) {
+      return [];
+    }
+
+    return Object.keys(pLevels)
+      .filter((pair: string) => {
+        return (
+          normalizedTokenAddress === pLevels[pair].base ||
+          normalizedTokenAddress === pLevels[pair].quote
+        );
+      })
+      .map((pair: string) => {
+        return {
+          exchange: this.dexKey,
+          address: this.routerAddressPlaceholder,
+          connectorTokens: [this.getTokenFromAddress(pLevels[pair].quote!)],
```

Comment:
I am not sure this is correct. From line 686 I see that `getTokenFromAddress` is mock implementation. It doesn't give real decimals. Here we expect correct data.

Another thing. Maybe I just miss something. Why we assume here that `quote` is not equal to normalizedTokenAddress? I mean, in line 667-668 we filter by both criteria:
```
          normalizedTokenAddress === pLevels[pair].base ||
          normalizedTokenAddress === pLevels[pair].quote
```
In this case I expect, we should have something like:
```
normalizedTokenAddress === pLevels[pair].base ? pLevels[pair].quote : pLevels[pair].base
```
Wdyt?